In [1]:
from datetime import datetime, timedelta
import pandas as pd 
import matplotlib.pyplot as plt 
import numpy as np 
import seaborn as sns 
import plotly.offline as pyoff
import plotly.graph_objs as go 

In [2]:
#df_retail = pd.read_csv('C:\\Users\\LU VINH XUYEN\\Downloads\\OnlineRetail.csv')
df_retail = pd.read_csv('C:\\Users\\LU VINH XUYEN\\Downloads\\onlineretail.csv',  encoding='latin-1')
df_retail.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

In [3]:
df_retail.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


In [4]:
#Convert InvoiceDate to 'datetime'
df_retail['InvoiceDate'] = pd.to_datetime(df_retail['InvoiceDate'])
df_retail.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


In [5]:
#Extract year & month
df_retail['year'] = df_retail['InvoiceDate'].dt.year
df_retail['month'] = df_retail['InvoiceDate'].dt.month
df_retail['InvoiceYearMonth'] = df_retail['InvoiceDate'].dt.strftime('%Y%m')

#df_retail.head(5)

In [6]:
df_retail.head(5)

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  year  month  \
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  2010     12   
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010     12   
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  2010     12   
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010     12   
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010     12   

  InvoiceYearMonth  
0           201012  
1           201012  
2           201012  
3           201012  
4           201012

In [7]:
#Calculate revenue = unitprice * quantity
df_retail['Revenue'] = df_retail['UnitPrice']* df_retail['Quantity']

Monthly Growth rate by revenue ?

In [8]:
df_revenue = df_retail.groupby('InvoiceYearMonth')['Revenue'].sum().reset_index()

In [9]:
pct_change_list = [0] #If only run loop, the result has only 12 values because pct_change start calculate ion the second value. So must make a '0' value to match with df_revenue
for i in range(len(df_revenue)-1): #Because this df has 13 rows but start at 0
    pct_change = (df_revenue['Revenue'][i+1] - df_revenue['Revenue'][i])/df_revenue['Revenue'][i]
    pct_change_list.append(pct_change)
print(pct_change_list)

[0, -0.252293195676302, -0.110602823648689, 0.37184966590046437, -0.278163495012814, 0.4665917810217506, -0.04453048221144906, -0.014213110104028877, 0.002026124724937812, 0.4936527512701365, 0.05003203618372446, 0.36522823796033327, -0.7033239912605128]


In [10]:
df_revenue['pct_change'] = pct_change_list

In [11]:
plot_data = [
    go.Scatter(
        x= df_revenue[df_revenue['InvoiceYearMonth'] !='201112']['InvoiceYearMonth'],
        y= df_revenue[df_revenue['pct_change'] != '201112']['pct_change'],)
]    

plot_layout = go.Layout(
    xaxis= {"type":"category"},
    title = 'Monthly Growth Rate')

fig = go.Figure(data=plot_data, layout = plot_layout)
pyoff.plot(fig)


'temp-plot.html'

Monthly Active Customers

In [12]:
df_retail['Country'].value_counts()

United Kingdom          495478
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
USA                        291
Hong Kong                  288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58
Lebanon 

United Kingdom has overwhelming numbers of active, so we focus on this country

In [13]:
df_uk = df_retail[df_retail['Country'] =='United Kingdom'].reset_index(drop = True) # can't using inplace

In [14]:
df_monthly_active = df_uk.groupby('InvoiceYearMonth')['CustomerID'].nunique().reset_index()
df_monthly_active

InvoiceYearMonth  CustomerID
0            201012         871
1            201101         684
2            201102         714
3            201103         923
4            201104         817
5            201105         985
6            201106         943
7            201107         899
8            201108         867
9            201109        1177
10           201110        1285
11           201111        1548
12           201112         617

In [15]:
plot_data = [
    go.Bar(
        x= df_monthly_active['InvoiceYearMonth'],
        y= df_monthly_active['CustomerID'],
                             ) 
]    

plot_layout = go.Layout(
    xaxis= {"type":"category"},
    title = 'Monthly Active Customers')

fig = go.Figure(data=plot_data, layout = plot_layout)
pyoff.plot(fig)


'temp-plot.html'

What is average of products sold and revenue by month ? 

In [16]:
df_monthly_sales = df_uk.groupby('InvoiceYearMonth')['Quantity'].sum().reset_index()
df_monthly_sales

InvoiceYearMonth  Quantity
0            201012    298101
1            201101    237381
2            201102    225641
3            201103    279843
4            201104    257666
5            201105    306452
6            201106    258522
7            201107    324129
8            201108    319804
9            201109    458490
10           201110    455612
11           201111    642281
12           201112    199907

In [17]:
plot_data = [
    go.Scatter(
        x= df_monthly_sales['InvoiceYearMonth'] ,
        y= df_monthly_sales['Quantity'] ,)
]    

plot_layout = go.Layout(
    xaxis= {"type":"category"},
    title = 'Monthly Sold Products')

fig = go.Figure(data=plot_data, layout = plot_layout)
pyoff.plot(fig)

'temp-plot.html'

In [18]:
mean_quantity = df_monthly_sales['Quantity'].mean()

In [19]:
mean_quantity = round(mean_quantity,2)
mean_quantity

327986.85

Above is the result for mean of sold products 

In [20]:
df_monthly_order_avg = df_uk.groupby('InvoiceYearMonth')['Revenue'].mean().reset_index() 
df_monthly_order_avg

InvoiceYearMonth    Revenue
0            201012  16.865860
1            201101  13.614680
2            201102  16.093027
3            201103  16.716166
4            201104  15.773380
5            201105  17.713823
6            201106  16.714748
7            201107  15.723497
8            201108  17.315899
9            201109  18.931723
10           201110  16.093582
11           201111  16.312383
12           201112  16.247406

Above is the result for the average of revenue by month

Find New & Existing Customers

In [21]:
#Find minimum of Invoice date for each customers
df_min_purchase = df_uk.groupby('CustomerID').InvoiceDate.min().reset_index()

In [22]:
df_min_purchase.columns= ['CustomerID' , 'MinPurchase']
df_min_purchase

CustomerID         MinPurchase
0        12346.0 2011-01-18 10:01:00
1        12747.0 2010-12-05 15:38:00
2        12748.0 2010-12-01 12:48:00
3        12749.0 2011-05-10 15:25:00
4        12820.0 2011-01-17 12:34:00
...          ...                 ...
3945     18280.0 2011-03-07 09:52:00
3946     18281.0 2011-06-12 10:53:00
3947     18282.0 2011-08-05 13:35:00
3948     18283.0 2011-01-06 14:14:00
3949     18287.0 2011-05-22 10:39:00

[3950 rows x 2 columns]

Check

In [23]:
df_min_purchase['MinPurchasedYearMonth'] = df_min_purchase['MinPurchase'].map(lambda x: x.year*100 + x.month) #Other way to extract & concat year - month

In [24]:
#Join df_min_purchase with df_uk
df_uk = pd.merge(df_uk, df_min_purchase, on ='CustomerID') 
df_uk 

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
361873    581416     22809              SET OF 6 T-LIGHTS SANTA         1   
361874    581416     22807         SET OF 6 T-LIGHTS TOADSTOOLS         2   
361875    581416    72349B      SET/6 PURPLE BUTTERFLY T-LIGHTS         1   
361876    581416     22809              SET OF 6 T-LIGHTS SANTA         2   
361877    581416     23487             SWEET HEART CAKE CARRIER         1   

               InvoiceDate  UnitPrice  CustomerID         Country  year  \
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  2010   
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010   
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  2010   
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010   
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010   
...                    ...        ...         ...             ...   ...   
361873 2011-12-08 14:58:00       2.95     14569.0  United Kingdom  2011   
361874 2011-12-08 14:58:00       1.25     14569.0  United Kingdom  2011   
361875 2011-12-08 14:58:00       2.10     14569.0  United Kingdom  2011   
361876 2011-12-08 14:58:00       2.95     14569.0  United Kingdom  2011   
361877 2011-12-08 14:58:00       9.95     14569.0  United Kingdom  2011   

        month InvoiceYearMonth  Revenue         MinPurchase  \
0          12           201012    15.30 2010-12-01 08:26:00   
1          12           201012    20.34 2010-12-01 08:26:00   
2          12           201012    22.00 2010-12-01 08:26:00   
3          12           201012    20.34 2010-12-01 08:26:00   
4          12           201012    20.34 2010-12-01 08:26:00   
...       ...              ...      ...                 ...   
361873     12           201112     2.95 2011-12-08 14:58:00   
361874     12           201112     2.50 2011-12-08 14:58:00   
361875     12           201112     2.10 2011-12-08 14:58:00   
361876     12           201112     5.90 2011-12-08 14:58:00   
361877     12           201112     9.95 2011-12-08 14:58:00   

        MinPurchasedYearMonth  
0                      201012  
1                      201012  
2                      201012  
3                      201012  
4                      201012  
...                       ...  
361873                 201112  
361874                 201112  
361875                 201112  
361876                 201112  
361877                 201112  

[361878 rows x 14 columns]

In [25]:
# Define all remaining customers is 'existing'
#df_uk.loc[df_uk['InvoiceYearMonth'] = df_uk['MinPurchasedYearMonth'],'UserType' ] = 'New' 
# Define all remaining customers is 'existing'
#df_uk.loc[df_uk['InvoiceYearMonth'] != df_uk['MinPurchasedYearMonth'],'UserType' ] = 'Existing' 
df_uk['UserType'] = None

In [26]:
#df_uk.loc[df_uk['InvoiceYearMonth'] == df_uk['MinPurchasedYearMonth'],'UserType' ] = 'New' 
#df_uk.loc[df_uk['InvoiceYearMonth'] != df_uk['MinPurchasedYearMonth'],'UserType' ] = 'Existing' 
# failed, try if else

In [27]:
df_uk.info()
# I saw the format of 2 columns that try to compare is not match, so must convert 1 of 2

<class 'pandas.core.frame.DataFrame'>
Int64Index: 361878 entries, 0 to 361877
Data columns (total 15 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   InvoiceNo              361878 non-null  object        
 1   StockCode              361878 non-null  object        
 2   Description            361878 non-null  object        
 3   Quantity               361878 non-null  int64         
 4   InvoiceDate            361878 non-null  datetime64[ns]
 5   UnitPrice              361878 non-null  float64       
 6   CustomerID             361878 non-null  float64       
 7   Country                361878 non-null  object        
 8   year                   361878 non-null  int64         
 9   month                  361878 non-null  int64         
 10  InvoiceYearMonth       361878 non-null  object        
 11  Revenue                361878 non-null  float64       
 12  MinPurchase            361878 non-null  date

Check format

In [28]:
df_uk['InvoiceYearMonth'] = pd.to_numeric(df_retail['InvoiceYearMonth'])


In [29]:
df_uk.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 361878 entries, 0 to 361877
Data columns (total 15 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   InvoiceNo              361878 non-null  object        
 1   StockCode              361878 non-null  object        
 2   Description            361878 non-null  object        
 3   Quantity               361878 non-null  int64         
 4   InvoiceDate            361878 non-null  datetime64[ns]
 5   UnitPrice              361878 non-null  float64       
 6   CustomerID             361878 non-null  float64       
 7   Country                361878 non-null  object        
 8   year                   361878 non-null  int64         
 9   month                  361878 non-null  int64         
 10  InvoiceYearMonth       361878 non-null  int64         
 11  Revenue                361878 non-null  float64       
 12  MinPurchase            361878 non-null  date

In [32]:
def compare_and_label(row):
  if row['InvoiceYearMonth'] == row['MinPurchasedYearMonth']:
    return 'New'
  else:
    return 'Existing'

# Apply the function to each row and create a new column
df_uk['New_Existing'] = df_uk.apply(compare_and_label, axis=1)

print(df_uk)

       InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
361873    581416     22809              SET OF 6 T-LIGHTS SANTA         1   
361874    581416     22807         SET OF 6 T-LIGHTS TOADSTOOLS         2   
361875    581416    72349B      SET/6 PURPLE BUTTERFLY T-LIGHTS         1   
361876    581416     22809              SET OF 6 T-LIGHTS SANTA         2   
361877    581416     23487             SWEET HEART CAKE CARRIER         1   

               InvoiceDate  UnitPrice  CustomerID         Country  year  \


In [33]:
df_uk['New_Existing'].value_counts()

Existing    295058
New          66820
Name: New_Existing, dtype: int64

In [35]:
#Calculate revenue of each user_type 
df_uk = df_uk.drop('UserType', axis=1)


In [36]:
df_uk.head(5)

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  year  month  \
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  2010     12   
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010     12   
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  2010     12   
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010     12   
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010     12   

   InvoiceYearMonth  Revenue         MinPurchase  MinPurchasedYearMonth  \
0            201012    15.30 2010-12-01 08:26:00                 201012   
1            201012    20.34 2010-12-01 08:26:00                 201012   
2            201012    22.00 2010-12-01 08:26:00                 201012   
3            201012    20.34 2010-12-01 08:26:00                 201012   
4            201012    20.34 2010-12-01 08:26:00                 201012   

  New_Existing  
0          New  
1          New  
2          New  
3          New  
4          New

In [37]:
df_uk = df_uk.rename(columns={'New_Exisiting': 'UserType'})


In [41]:
df_uk = df_uk.rename(columns={'New_Exisiting': 'UserType'})
df_uk.head(5)

New_Exisiting InvoiceNo StockCode                          Description  \
0                536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER   
1                536365     71053                  WHITE METAL LANTERN   
2                536365    84406B       CREAM CUPID HEARTS COAT HANGER   
3                536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE   
4                536365    84029E       RED WOOLLY HOTTIE WHITE HEART.   

New_Exisiting  Quantity         InvoiceDate  UnitPrice  CustomerID  \
0                     6 2010-12-01 08:26:00       2.55     17850.0   
1                     6 2010-12-01 08:26:00       3.39     17850.0   
2                     8 2010-12-01 08:26:00       2.75     17850.0   
3                     6 2010-12-01 08:26:00       3.39     17850.0   
4                     6 2010-12-01 08:26:00       3.39     17850.0   

New_Exisiting         Country  year  month  InvoiceYearMonth  Revenue  \
0              United Kingdom  2010     12            201012    15.30   
1              United Kingdom  2010     12            201012    20.34   
2              United Kingdom  2010     12            201012    22.00   
3              United Kingdom  2010     12            201012    20.34   
4              United Kingdom  2010     12            201012    20.34   

New_Exisiting         MinPurchase  MinPurchasedYearMonth New_Existing  
0             2010-12-01 08:26:00                 201012          New  
1             2010-12-01 08:26:00                 201012          New  
2             2010-12-01 08:26:00                 201012          New  
3             2010-12-01 08:26:00                 201012          New  
4             2010-12-01 08:26:00                 201012          New

In [42]:
df_user_type_revenue = df_uk.groupby(['InvoiceYearMonth','New_Existing'])['Revenue'].sum().reset_index()

In [43]:
df_user_type_revenue

InvoiceYearMonth New_Existing     Revenue
0             201012          New  924245.940
1             201101     Existing  722621.050
2             201102     Existing  932153.880
3             201103     Existing  607561.740
4             201104     Existing  645326.760
5             201105     Existing  627999.431
6             201106     Existing  644078.910
7             201107     Existing  638544.841
8             201108     Existing  330516.121
9             201108          New  147302.910
10            201109     Existing  343963.520
11            201109          New  203558.291

In [44]:
df_user_type_revenue = df_user_type_revenue.query("InvoiceYearMonth != 201012 and InvoiceYearMonth!= 201112")
#Exclude first and last month

In [46]:
plot_data = [
    go.Scatter(
    x=df_user_type_revenue.query("New_Existing == 'Existing'")['InvoiceYearMonth'],
    y=df_user_type_revenue.query("New_Existing == 'Existing'")['Revenue'],
    name = 'Existing'),
    go.Scatter(
    x=df_user_type_revenue.query("New_Existing == 'New'")['InvoiceYearMonth'],
    y=df_user_type_revenue.query("New_Existing == 'New'")['Revenue'],
    name = 'New')
]

plot_layout = go.Layout(
    xaxis={"type":"category"},
    title='Revenue of New vs Existing Customers')

fig = go.Figure(data = plot_data, layout = plot_layout)
pyoff.plot(fig)

'temp-plot.html'

Monthly New Customers Ratio

In [47]:
countNew = df_uk.query("New_Existing == 'New'").groupby(['InvoiceYearMonth'])['CustomerID'].nunique()

Count New customers

In [48]:
countNew

InvoiceYearMonth
201012    177
201108    133
201109    275
Name: CustomerID, dtype: int64

In [49]:
countExisting = df_uk.query("New_Existing == 'Existing'").groupby(['InvoiceYearMonth'])['CustomerID'].nunique()

In [50]:
df_user_ratio = countNew/countExisting

In [51]:
df_user_ratio

InvoiceYearMonth
201012         NaN
201101         NaN
201102         NaN
201103         NaN
201104         NaN
201105         NaN
201106         NaN
201107         NaN
201108    0.288503
201109    0.419207
Name: CustomerID, dtype: float64

In [52]:
df_user_ratiov = df_user_ratio.dropna()

In [54]:
df_user_ratiov

InvoiceYearMonth
201108    0.288503
201109    0.419207
Name: CustomerID, dtype: float64

In [56]:
df_user_ratiov.columns= ['InvoiceYearMonth' , 'Ratio']
df_user_ratiov

InvoiceYearMonth
201108    0.288503
201109    0.419207
Name: CustomerID, dtype: float64

In [57]:
df_user_ratiov = df_user_ratiov.to_frame()

In [60]:
df_user_ratiov = df_user_ratiov.reset_index()
df_user_ratiov

InvoiceYearMonth  CustomerID
0            201108    0.288503
1            201109    0.419207

In [61]:
plot_data = [
    go.Bar(
        x=df_user_ratiov['InvoiceYearMonth'],
        y=df_user_ratiov['CustomerID'],)
]

plot_layout = go.Layout(
    xaxis = {"type" :"category"},
    title = 'New Customers Ratio')

fig = go.Figure(data = plot_data, layout = plot_layout)
pyoff.plot(fig)

'temp-plot.html'

Monthly Retention Rate 

In [62]:
df_uk['InvoiceYearMonth'] = df_uk['InvoiceDate'].map(lambda date: 100*date.year + date.month)

In [63]:
df_uk

New_Exisiting InvoiceNo StockCode                          Description  \
0                536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER   
1                536365     71053                  WHITE METAL LANTERN   
2                536365    84406B       CREAM CUPID HEARTS COAT HANGER   
3                536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE   
4                536365    84029E       RED WOOLLY HOTTIE WHITE HEART.   
...                 ...       ...                                  ...   
361873           581416     22809              SET OF 6 T-LIGHTS SANTA   
361874           581416     22807         SET OF 6 T-LIGHTS TOADSTOOLS   
361875           581416    72349B      SET/6 PURPLE BUTTERFLY T-LIGHTS   
361876           581416     22809              SET OF 6 T-LIGHTS SANTA   
361877           581416     23487             SWEET HEART CAKE CARRIER   

New_Exisiting  Quantity         InvoiceDate  UnitPrice  CustomerID  \
0                     6 2010-12-01 08:26:00       2.55     17850.0   
1                     6 2010-12-01 08:26:00       3.39     17850.0   
2                     8 2010-12-01 08:26:00       2.75     17850.0   
3                     6 2010-12-01 08:26:00       3.39     17850.0   
4                     6 2010-12-01 08:26:00       3.39     17850.0   
...                 ...                 ...        ...         ...   
361873                1 2011-12-08 14:58:00       2.95     14569.0   
361874                2 2011-12-08 14:58:00       1.25     14569.0   
361875                1 2011-12-08 14:58:00       2.10     14569.0   
361876                2 2011-12-08 14:58:00       2.95     14569.0   
361877                1 2011-12-08 14:58:00       9.95     14569.0   

New_Exisiting         Country  year  month  InvoiceYearMonth  Revenue  \
0              United Kingdom  2010     12            201012    15.30   
1              United Kingdom  2010     12            201012    20.34   
2              United Kingdom  2010     12            201012    22.00   
3              United Kingdom  2010     12            201012    20.34   
4              United Kingdom  2010     12            201012    20.34   
...                       ...   ...    ...               ...      ...   
361873         United Kingdom  2011     12            201112     2.95   
361874         United Kingdom  2011     12            201112     2.50   
361875         United Kingdom  2011     12            201112     2.10   
361876         United Kingdom  2011     12            201112     5.90   
361877         United Kingdom  2011     12            201112     9.95   

New_Exisiting         MinPurchase  MinPurchasedYearMonth New_Existing  
0             2010-12-01 08:26:00                 201012          New  
1             2010-12-01 08:26:00                 201012          New  
2             2010-12-01 08:26:00                 201012          New  
3             2010-12-01 08:26:00                 201012          New  
4             2010-12-01 08:26:00                 201012          New  
...                           ...                    ...          ...  
361873        2011-12-08 14:58:00                 201112     Existing  
361874        2011-12-08 14:58:00                 201112     Existing  
361875        2011-12-08 14:58:00                 201112     Existing  
361876        2011-12-08 14:58:00                 201112     Existing  
361877        2011-12-08 14:58:00                 201112     Existing  

[361878 rows x 15 columns]

In [64]:
df_monthly_active = df_uk.groupby('InvoiceYearMonth')['CustomerID'].nunique().reset_index()

In [65]:
df_user_purchase = df_uk.groupby(['CustomerID','InvoiceYearMonth'])['Revenue'].sum().astype(int).reset_index()
df_user_purchase

CustomerID  InvoiceYearMonth  Revenue
0         12346.0            201101        0
1         12747.0            201012      706
2         12747.0            201101      303
3         12747.0            201103      310
4         12747.0            201105      771
...           ...               ...      ...
12325     18283.0            201110      114
12326     18283.0            201111      651
12327     18283.0            201112      208
12328     18287.0            201105      765
12329     18287.0            201110     1072

[12330 rows x 3 columns]

In [66]:
df_retention = pd.crosstab(df_user_purchase['CustomerID'], df_user_purchase['InvoiceYearMonth']).reset_index()

In [67]:
df_retention

InvoiceYearMonth  CustomerID  201012  201101  201102  201103  201104  201105  \
0                    12346.0       0       1       0       0       0       0   
1                    12747.0       1       1       0       1       0       1   
2                    12748.0       1       1       1       1       1       1   
3                    12749.0       0       0       0       0       0       1   
4                    12820.0       0       1       0       0       0       0   
...                      ...     ...     ...     ...     ...     ...     ...   
3945                 18280.0       0       0       0       1       0       0   
3946                 18281.0       0       0       0       0       0       0   
3947                 18282.0       0       0       0       0       0       0   
3948                 18283.0       0       1       1       0       1       1   
3949                 18287.0       0       0       0       0       0       1   

InvoiceYearMonth  201106  201107  201108  201109  201110  201111  201112  
0                      0       0       0       0       0       0       0  
1                      1       0       1       0       1       1       1  
2                      1       1       1       1       1       1       1  
3                      0       0       1       0       0       1       1  
4                      0       0       0       1       1       0       1  
...                  ...     ...     ...     ...     ...     ...     ...  
3945                   0       0       0       0       0       0       0  
3946                   1       0       0       0       0       0       0  
3947                   0       0       1       0       0       0       1  
3948                   1       1       0       1       1       1       1  
3949                   0       0       0       0       1       0       0  

[3950 rows x 14 columns]

In [68]:
months = df_retention.columns

In [69]:
months 

Index(['CustomerID',       201012,       201101,       201102,       201103,
             201104,       201105,       201106,       201107,       201108,
             201109,       201110,       201111,       201112],
      dtype='object', name='InvoiceYearMonth')

The table is resulted by customer_id and invoiceyearmonth, if value is 0, the customer didn't buy in this month

In [70]:
retention_array = []
for i in range(len(months)-1):
    retention_data = {}
    selected_month = months[i+1]
    prev_month = months[i] 
    retention_data['InvoiceYearMonth'] = int(selected_month)
    retention_data['TotalUserCount'] = df_retention[selected_month].sum()
    retention_data['RetainedUserCount'] = df_retention[(df_retention[selected_month]==1) & (df_retention[prev_month]==1)][selected_month].sum()
    retention_array.append(retention_data)
    print("*******" + str(selected_month) + "*******") 
    print(retention_array)
    #Make an array can show value by month, include previous month

*******201012*******
[{'InvoiceYearMonth': 201012, 'TotalUserCount': 871, 'RetainedUserCount': 0}]
*******201101*******
[{'InvoiceYearMonth': 201012, 'TotalUserCount': 871, 'RetainedUserCount': 0}, {'InvoiceYearMonth': 201101, 'TotalUserCount': 684, 'RetainedUserCount': 322}]
*******201102*******
[{'InvoiceYearMonth': 201012, 'TotalUserCount': 871, 'RetainedUserCount': 0}, {'InvoiceYearMonth': 201101, 'TotalUserCount': 684, 'RetainedUserCount': 322}, {'InvoiceYearMonth': 201102, 'TotalUserCount': 714, 'RetainedUserCount': 263}]
*******201103*******
[{'InvoiceYearMonth': 201012, 'TotalUserCount': 871, 'RetainedUserCount': 0}, {'InvoiceYearMonth': 201101, 'TotalUserCount': 684, 'RetainedUserCount': 322}, {'InvoiceYearMonth': 201102, 'TotalUserCount': 714, 'RetainedUserCount': 263}, {'InvoiceYearMonth': 201103, 'TotalUserCount': 923, 'RetainedUserCount': 305}]
*******201104*******
[{'InvoiceYearMonth': 201012, 'TotalUserCount': 871, 'RetainedUserCount': 0}, {'InvoiceYearMonth': 201101, 'T

In [71]:
df_retention = pd.DataFrame(retention_array)
df_retention

InvoiceYearMonth  TotalUserCount  RetainedUserCount
0             201012             871                  0
1             201101             684                322
2             201102             714                263
3             201103             923                305
4             201104             817                310
5             201105             985                369
6             201106             943                417
7             201107             899                379
8             201108             867                391
9             201109            1177                417
10            201110            1285                502
11            201111            1548                616
12            201112             617                402

In 201101, there are 322 customers in 201012 still had purchase. Similar to the remaining months

In [72]:
df_retention['RetentionRate'] = df_retention['RetainedUserCount']/df_retention['TotalUserCount']
df_retention

InvoiceYearMonth  TotalUserCount  RetainedUserCount  RetentionRate
0             201012             871                  0       0.000000
1             201101             684                322       0.470760
2             201102             714                263       0.368347
3             201103             923                305       0.330444
4             201104             817                310       0.379437
5             201105             985                369       0.374619
6             201106             943                417       0.442206
7             201107             899                379       0.421580
8             201108             867                391       0.450980
9             201109            1177                417       0.354291
10            201110            1285                502       0.390661
11            201111            1548                616       0.397933
12            201112             617                402       0.651540

In [73]:
plot_data = [
    go.Scatter(
    x= df_retention['InvoiceYearMonth'],
    y= df_retention['RetentionRate'],
    name='organic'
    )
]

plot_layout = go.Layout(
    xaxis = {"type":"category"},
    title = 'Monthly Retention Rate')
fig = go.Figure(data = plot_data, layout = plot_layout)
pyoff.plot(fig)

'temp-plot.html'

Churn rate

In [74]:
df_retention['ChurnRate'] = 1 - df_retention['RetentionRate']
df_retention

InvoiceYearMonth  TotalUserCount  RetainedUserCount  RetentionRate  \
0             201012             871                  0       0.000000   
1             201101             684                322       0.470760   
2             201102             714                263       0.368347   
3             201103             923                305       0.330444   
4             201104             817                310       0.379437   
5             201105             985                369       0.374619   
6             201106             943                417       0.442206   
7             201107             899                379       0.421580   
8             201108             867                391       0.450980   
9             201109            1177                417       0.354291   
10            201110            1285                502       0.390661   
11            201111            1548                616       0.397933   
12            201112             617                402       0.651540   

    ChurnRate  
0    1.000000  
1    0.529240  
2    0.631653  
3    0.669556  
4    0.620563  
5    0.625381  
6    0.557794  
7    0.578420  
8    0.549020  
9    0.645709  
10   0.609339  
11   0.602067  
12   0.348460

In [75]:
plot_data = [
    go.Scatter(
    x= df_retention.query("InvoiceYearMonth>201101")['InvoiceYearMonth'],
    y= df_retention.query("InvoiceYearMonth>201101")['ChurnRate'],
    name='organic'
    )
]

plot_layout = go.Layout(
    xaxis = {"type":"category"},
    title = 'Monthly Churn Rate')
fig = go.Figure(data = plot_data, layout = plot_layout)
pyoff.plot(fig)
#Exclude month 201012

'temp-plot.html'

Cohort Retention

In [78]:
df_min_purchase.head(5)

CustomerID         MinPurchase  MinPurchasedYearMonth
0     12346.0 2011-01-18 10:01:00                 201101
1     12747.0 2010-12-05 15:38:00                 201012
2     12748.0 2010-12-01 12:48:00                 201012
3     12749.0 2011-05-10 15:25:00                 201105
4     12820.0 2011-01-17 12:34:00                 201101

In [79]:
df_min_purchase['MinPurchasedYearMonth'] = df_min_purchase['MinPurchase'].map(lambda date: 100*date.year + date.month)

In [80]:
df_retention.head(5)

InvoiceYearMonth  TotalUserCount  RetainedUserCount  RetentionRate  \
0            201012             871                  0       0.000000   
1            201101             684                322       0.470760   
2            201102             714                263       0.368347   
3            201103             923                305       0.330444   
4            201104             817                310       0.379437   

   ChurnRate  
0   1.000000  
1   0.529240  
2   0.631653  
3   0.669556  
4   0.620563

In [81]:
df_retention = pd.crosstab(df_user_purchase['CustomerID'], df_user_purchase['InvoiceYearMonth']).reset_index()

CustomerID  201012  201101  201102  201103  201104  201105  201106  \
0        12346.0       0       1       0       0       0       0       0   
1        12747.0       1       1       0       1       0       1       1   
2        12748.0       1       1       1       1       1       1       1   
3        12749.0       0       0       0       0       0       1       0   
4        12820.0       0       1       0       0       0       0       0   
...          ...     ...     ...     ...     ...     ...     ...     ...   
3945     18280.0       0       0       0       1       0       0       0   
3946     18281.0       0       0       0       0       0       0       1   
3947     18282.0       0       0       0       0       0       0       0   
3948     18283.0       0       1       1       0       1       1       1   
3949     18287.0       0       0       0       0       0       1       0   

      201107  201108  201109  201110  201111  201112  MinPurchasedYearMonth  
0          0       0       0       0       0       0                 201101  
1          0       1       0       1       1       1                 201012  
2          1       1       1       1       1       1                 201012  
3          0       1       0       0       1       1                 201105  
4          0       0       1       1       0       1                 201101  
...      ...     ...     ...     ...     ...     ...                    ...  
3945       0       0       0       0       0       0                 201103  
3946       0       0       0       0       0       0                 201106  
3947       0       1       0       0       0       1                 201108  
3948       1       0       1       1       1       1                 201101  
3949       0       0       0       1       0       0                 201105  

[3950 rows x 15 columns]

In [84]:
new_column_names = ['m' + str(column) for column in df_retention.columns[:-1]] 
new_column_names.append('MinPurchasedYearMonth') 
df_retention.columns = new_column_names

Check

In [94]:
months = df_retention.columns[2:]
months

Index(['m201101', 'm201102', 'm201103', 'm201104', 'm201105', 'm201106',
       'm201107', 'm201108', 'm201109', 'm201110', 'm201111', 'm201112',
       'MinPurchasedYearMonth'],
      dtype='object')

In [86]:
df_retention.head(5)

mCustomerID  m201012  m201101  m201102  m201103  m201104  m201105  m201106  \
0      12346.0        0        1        0        0        0        0        0   
1      12747.0        1        1        0        1        0        1        1   
2      12748.0        1        1        1        1        1        1        1   
3      12749.0        0        0        0        0        0        1        0   
4      12820.0        0        1        0        0        0        0        0   

   m201107  m201108  m201109  m201110  m201111  m201112  MinPurchasedYearMonth  
0        0        0        0        0        0        0                 201101  
1        0        1        0        1        1        1                 201012  
2        1        1        1        1        1        1                 201012  
3        0        1        0        0        1        1                 201105  
4        0        0        1        1        0        1                 201101

In [87]:
total_user_count = df_retention[df_retention['MinPurchasedYearMonth'] ==201101 ]['MinPurchasedYearMonth'].count()

In [89]:
df_retention.query("MinPurchasedYearMonth ==201101 and {} ==1".format(str('m' +str(201102))))['m' +str(201102)].sum()/total_user_count

0.23204419889502761

In [93]:
months = df_retention.columns[2:]
months

Index(['m201101', 'm201102', 'm201103', 'm201104', 'm201105', 'm201106',
       'm201107', 'm201108', 'm201109', 'm201110', 'm201111', 'm201112',
       'MinPurchasedYearMonth'],
      dtype='object')

In [95]:
months = [element[1:] for element in months] 
months

['201101',
 '201102',
 '201103',
 '201104',
 '201105',
 '201106',
 '201107',
 '201108',
 '201109',
 '201110',
 '201111',
 '201112',
 'inPurchasedYearMonth']

In [96]:
retention_array = []
for i in range(len(months)):
    retention_data = {}
    selected_month = months[i]
    prev_months = months[:i]
    next_months = months[i+1:]
    for prev_month in prev_months:
        retention_data[prev_month] = np.nan #All values of prev month is NaN
        
    total_user_count = df_retention[df_retention.MinPurchasedYearMonth == selected_month].MinPurchasedYearMonth.count()
    retention_data['TotalUserCount'] = total_user_count 
    retention_data[selected_month] =1
    
    query = "MinPurchasedYearMonth == {}".format(selected_month) 
    
    for next_month in next_months:
        new_query = query + "and {} >0".format(str('m' +str(next_month)))
        retention_data[next_month] = np.round(df_retention.query(new_query)['m' + str(next_month)].sum()/total_user_count,2)
    retention_array.append(retention_data)
    
df_retention = pd.DataFrame(retention_array)

C:\Users\LU VINH XUYEN\AppData\Local\Temp\ipykernel_11928\4207074358.py:18: RuntimeWarning:

divide by zero encountered in longlong_scalars

C:\Users\LU VINH XUYEN\AppData\Local\Temp\ipykernel_11928\4207074358.py:18: RuntimeWarning:

divide by zero encountered in longlong_scalars

C:\Users\LU VINH XUYEN\AppData\Local\Temp\ipykernel_11928\4207074358.py:18: RuntimeWarning:

divide by zero encountered in longlong_scalars

C:\Users\LU VINH XUYEN\AppData\Local\Temp\ipykernel_11928\4207074358.py:18: RuntimeWarning:

divide by zero encountered in longlong_scalars

C:\Users\LU VINH XUYEN\AppData\Local\Temp\ipykernel_11928\4207074358.py:18: RuntimeWarning:

divide by zero encountered in longlong_scalars

C:\Users\LU VINH XUYEN\AppData\Local\Temp\ipykernel_11928\4207074358.py:18: RuntimeWarning:

divide by zero encountered in longlong_scalars

C:\Users\LU VINH XUYEN\AppData\Local\Temp\ipykernel_11928\4207074358.py:18: RuntimeWarning:

divide by zero encountered in longlong_scalars

C:\Users\LU V

UndefinedVariableError: name 'minPurchasedYearMonth' is not defined

In [97]:
df_min_purchase.head(5)

CustomerID         MinPurchase  MinPurchasedYearMonth
0     12346.0 2011-01-18 10:01:00                 201101
1     12747.0 2010-12-05 15:38:00                 201012
2     12748.0 2010-12-01 12:48:00                 201012
3     12749.0 2011-05-10 15:25:00                 201105
4     12820.0 2011-01-17 12:34:00                 201101

In [98]:
df_retention

mCustomerID  m201012  m201101  m201102  m201103  m201104  m201105  \
0         12346.0        0        1        0        0        0        0   
1         12747.0        1        1        0        1        0        1   
2         12748.0        1        1        1        1        1        1   
3         12749.0        0        0        0        0        0        1   
4         12820.0        0        1        0        0        0        0   
...           ...      ...      ...      ...      ...      ...      ...   
3945      18280.0        0        0        0        1        0        0   
3946      18281.0        0        0        0        0        0        0   
3947      18282.0        0        0        0        0        0        0   
3948      18283.0        0        1        1        0        1        1   
3949      18287.0        0        0        0        0        0        1   

      m201106  m201107  m201108  m201109  m201110  m201111  m201112  \
0           0        0        0        0        0        0        0   
1           1        0        1        0        1        1        1   
2           1        1        1        1        1        1        1   
3           0        0        1        0        0        1        1   
4           0        0        0        1        1        0        1   
...       ...      ...      ...      ...      ...      ...      ...   
3945        0        0        0        0        0        0        0   
3946        1        0        0        0        0        0        0   
3947        0        0        1        0        0        0        1   
3948        1        1        0        1        1        1        1   
3949        0        0        0        0        1        0        0   

      MinPurchasedYearMonth  
0                    201101  
1                    201012  
2                    201012  
3                    201105  
4                    201101  
...                     ...  
3945                 201103  
3946                 201106  
3947                 201108  
3948                 201101  
3949                 201105  

[3950 rows x 15 columns]

In [99]:
new_column_name = [element[1:] for element in df_retention.columns[:-1]]
new_column_name.append('MinPurchasedYearMonth') 
df_retention.columns = new_column_name

In [100]:
df_retention

CustomerID  201012  201101  201102  201103  201104  201105  201106  \
0        12346.0       0       1       0       0       0       0       0   
1        12747.0       1       1       0       1       0       1       1   
2        12748.0       1       1       1       1       1       1       1   
3        12749.0       0       0       0       0       0       1       0   
4        12820.0       0       1       0       0       0       0       0   
...          ...     ...     ...     ...     ...     ...     ...     ...   
3945     18280.0       0       0       0       1       0       0       0   
3946     18281.0       0       0       0       0       0       0       1   
3947     18282.0       0       0       0       0       0       0       0   
3948     18283.0       0       1       1       0       1       1       1   
3949     18287.0       0       0       0       0       0       1       0   

      201107  201108  201109  201110  201111  201112  MinPurchasedYearMonth  
0          0       0       0       0       0       0                 201101  
1          0       1       0       1       1       1                 201012  
2          1       1       1       1       1       1                 201012  
3          0       1       0       0       1       1                 201105  
4          0       0       1       1       0       1                 201101  
...      ...     ...     ...     ...     ...     ...                    ...  
3945       0       0       0       0       0       0                 201103  
3946       0       0       0       0       0       0                 201106  
3947       0       1       0       0       0       1                 201108  
3948       1       0       1       1       1       1                 201101  
3949       0       0       0       1       0       0                 201105  

[3950 rows x 15 columns]

In [104]:
month = df_retention.columns[2:]
month

Index(['201101', '201102', '201103', '201104', '201105', '201106', '201107',
       '201108', '201109', '201110', '201111', '201112',
       'MinPurchasedYearMonth'],
      dtype='object')

In [105]:
month[:3]

Index(['201101', '201102', '201103'], dtype='object')

In [107]:
df_retention

CustomerID  201012  201101  201102  201103  201104  201105  201106  \
0        12346.0       0       1       0       0       0       0       0   
1        12747.0       1       1       0       1       0       1       1   
2        12748.0       1       1       1       1       1       1       1   
3        12749.0       0       0       0       0       0       1       0   
4        12820.0       0       1       0       0       0       0       0   
...          ...     ...     ...     ...     ...     ...     ...     ...   
3945     18280.0       0       0       0       1       0       0       0   
3946     18281.0       0       0       0       0       0       0       1   
3947     18282.0       0       0       0       0       0       0       0   
3948     18283.0       0       1       1       0       1       1       1   
3949     18287.0       0       0       0       0       0       1       0   

      201107  201108  201109  201110  201111  201112  MinPurchasedYearMonth  
0          0       0       0       0       0       0                 201101  
1          0       1       0       1       1       1                 201012  
2          1       1       1       1       1       1                 201012  
3          0       1       0       0       1       1                 201105  
4          0       0       1       1       0       1                 201101  
...      ...     ...     ...     ...     ...     ...                    ...  
3945       0       0       0       0       0       0                 201103  
3946       0       0       0       0       0       0                 201106  
3947       0       1       0       0       0       1                 201108  
3948       1       0       1       1       1       1                 201101  
3949       0       0       0       1       0       0                 201105  

[3950 rows x 15 columns]

In [115]:
df_retention['201102'].sum()

714

In [116]:
total_user_count = df_retention[df_retention['MinPurchasedYearMonth'] ==201101 ]['MinPurchasedYearMonth'].count()

In [120]:
new_column_names = ['m' + str(column) for column in df_retention.columns[:-1]] 
new_column_names.append('MinPurchasedYearMonth') 
df_retention.columns = new_column_names

In [121]:
months

['201101',
 '201102',
 '201103',
 '201104',
 '201105',
 '201106',
 '201107',
 '201108',
 '201109',
 '201110',
 '201111',
 '201112',
 'inPurchasedYearMonth']

In [122]:
df_retention

mCustomerID  m201012  m201101  m201102  m201103  m201104  m201105  \
0         12346.0        0        1        0        0        0        0   
1         12747.0        1        1        0        1        0        1   
2         12748.0        1        1        1        1        1        1   
3         12749.0        0        0        0        0        0        1   
4         12820.0        0        1        0        0        0        0   
...           ...      ...      ...      ...      ...      ...      ...   
3945      18280.0        0        0        0        1        0        0   
3946      18281.0        0        0        0        0        0        0   
3947      18282.0        0        0        0        0        0        0   
3948      18283.0        0        1        1        0        1        1   
3949      18287.0        0        0        0        0        0        1   

      m201106  m201107  m201108  m201109  m201110  m201111  m201112  \
0           0        0        0        0        0        0        0   
1           1        0        1        0        1        1        1   
2           1        1        1        1        1        1        1   
3           0        0        1        0        0        1        1   
4           0        0        0        1        1        0        1   
...       ...      ...      ...      ...      ...      ...      ...   
3945        0        0        0        0        0        0        0   
3946        1        0        0        0        0        0        0   
3947        0        0        1        0        0        0        1   
3948        1        1        0        1        1        1        1   
3949        0        0        0        0        1        0        0   

      MinPurchasedYearMonth  
0                    201101  
1                    201012  
2                    201012  
3                    201105  
4                    201101  
...                     ...  
3945                 201103  
3946                 201106  
3947                 201108  
3948                 201101  
3949                 201105  

[3950 rows x 15 columns]

In [123]:
months = months[:-1]
months

['201101',
 '201102',
 '201103',
 '201104',
 '201105',
 '201106',
 '201107',
 '201108',
 '201109',
 '201110',
 '201111',
 '201112']

In [124]:
df_retention.query("MinPurchasedYearMonth ==201101 and {} ==1".format(str('m' +str(201102))))['m' +str(201102)].sum()/total_user_count

0.23204419889502761

In [125]:
months

['201101',
 '201102',
 '201103',
 '201104',
 '201105',
 '201106',
 '201107',
 '201108',
 '201109',
 '201110',
 '201111',
 '201112']

In [129]:
df_retention

TotalUserCount  201101  201102  201103  201104  201105  201106  201107  \
0                0     1.0     inf     inf     inf     inf     inf     inf   
1                0     NaN     1.0     inf     inf     inf     inf     inf   
2                0     NaN     NaN     1.0     inf     inf     inf     inf   
3                0     NaN     NaN     NaN     1.0     inf     inf     inf   
4                0     NaN     NaN     NaN     NaN     1.0     inf     inf   
5                0     NaN     NaN     NaN     NaN     NaN     1.0     inf   
6                0     NaN     NaN     NaN     NaN     NaN     NaN     1.0   
7                0     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
8                0     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
9                0     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
10               0     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

    201108  201109  201110  201111  201112  
0      inf     inf     inf     inf     inf  
1      inf     inf     inf     inf     inf  
2      inf     inf     inf     inf     inf  
3      inf     inf     inf     inf     inf  
4      inf     inf     inf     inf     inf  
5      inf     inf     inf     inf     inf  
6      inf     inf     inf     inf     inf  
7      1.0     inf     inf     inf     inf  
8      NaN     1.0     inf     inf     inf  
9      NaN     NaN     1.0     inf     inf  
10     NaN     NaN     NaN     1.0     inf

In [128]:
df_retention.query("MinPurchasedYearMonth ==201101 and {} ==1".format(str('m' +str(201103))))['m' +str(201103)].sum()/total_user_count

UndefinedVariableError: name 'MinPurchasedYearMonth' is not defined

In [126]:
retention_array = []
for i in range(len(months)-1):
    retention_data = {}
    selected_month = months[i]
    prev_months = months[:i]
    next_months = months[i+1:]
    for prev_month in prev_months:
        retention_data[prev_month] = np.nan #All values of prev month is NaN
        
    total_user_count = df_retention[df_retention.MinPurchasedYearMonth == selected_month].MinPurchasedYearMonth.count()
    retention_data['TotalUserCount'] = total_user_count 
    retention_data[selected_month] =1
    
    query = "MinPurchasedYearMonth == {}".format(selected_month) 
    
    for next_month in next_months:
        new_query = query + "and {} >0".format(str('m' +str(next_month)))
        retention_data[next_month] = np.round(df_retention.query(new_query)['m' + str(next_month)].sum()/total_user_count,2)
    retention_array.append(retention_data)
    
df_retention = pd.DataFrame(retention_array)

C:\Users\LU VINH XUYEN\AppData\Local\Temp\ipykernel_11928\693770892.py:18: RuntimeWarning:

divide by zero encountered in longlong_scalars

C:\Users\LU VINH XUYEN\AppData\Local\Temp\ipykernel_11928\693770892.py:18: RuntimeWarning:

divide by zero encountered in longlong_scalars

C:\Users\LU VINH XUYEN\AppData\Local\Temp\ipykernel_11928\693770892.py:18: RuntimeWarning:

divide by zero encountered in longlong_scalars

C:\Users\LU VINH XUYEN\AppData\Local\Temp\ipykernel_11928\693770892.py:18: RuntimeWarning:

divide by zero encountered in longlong_scalars

C:\Users\LU VINH XUYEN\AppData\Local\Temp\ipykernel_11928\693770892.py:18: RuntimeWarning:

divide by zero encountered in longlong_scalars

C:\Users\LU VINH XUYEN\AppData\Local\Temp\ipykernel_11928\693770892.py:18: RuntimeWarning:

divide by zero encountered in longlong_scalars

C:\Users\LU VINH XUYEN\AppData\Local\Temp\ipykernel_11928\693770892.py:18: RuntimeWarning:

divide by zero encountered in longlong_scalars

C:\Users\LU VINH XUY